src/array/cuda/csr_sum.cu

```diff
@@ -112,41 +114,52 @@ std::pair<CSRMatrix, NDArray> CSRSum(
 
   // Cast 64 bit indices to 32 bit
   std::vector<CSRMatrix> newAs;
+  newAs.reserve(n);
   bool cast = false;
   if (As[0].indptr->dtype.bits == 64) {
-    newAs.reserve(n);
     for (int i = 0; i < n; ++i)
       newAs.emplace_back(
         As[i].num_rows, As[i].num_cols, AsNumBits(As[i].indptr, 32),
         AsNumBits(As[i].indices, 32), AsNumBits(As[i].data, 32));
     cast = true;
+  } else {
+    for (int i = 0; i < n; ++i)
+      newAs.push_back(As[i]);
+  }
+
+  // cuSPARSE csrgeam2 requires the CSR to be sorted.
+  // TODO(BarclayII): ideally the sorted CSR should be cached but I'm not sure how to do it.
+  for (int i = 0; i < n; ++i) {
+    if (!newAs[i].sorted)
+      newAs[i] = CSRSort(newAs[i]);
   }
-  const std::vector<CSRMatrix> &As_ref = cast ? newAs : As;
 
   // Reorder weights if A[i] has edge IDs
   std::vector<NDArray> A_weights_reordered(n);
   for (int i = 0; i < n; ++i) {
-    if (CSRHasData(As[i]))
-      A_weights_reordered[i] = IndexSelect(A_weights[i], As[i].data);
+    if (CSRHasData(newAs[i]))
+      A_weights_reordered[i] = IndexSelect(A_weights[i], newAs[i].data);
     else
       A_weights_reordered[i] = A_weights[i];
   }
 
   // Loop and sum
   auto result = std::make_pair(
       CSRMatrix(
-        As_ref[0].num_rows, As_ref[0].num_cols,
-        As_ref[0].indptr, As_ref[0].indices),
+        newAs[0].num_rows, newAs[0].num_cols,
+        newAs[0].indptr, newAs[0].indices,
+        NullArray(newAs[0].indptr->dtype, newAs[0].indptr->ctx)),
```

Comment:
Found this in many places. Could we make it as part of the default behavior of `CSRMatrix` constructor?